turbopack/crates/turbopack-core/src/module_graph/side_effect_module_info.rs

```diff
@@ -0,0 +1,119 @@
+) -> Result<Vc<SideEffectFreeModules>> {
+    let parent_side_effect_free_modules = parent_side_effect_free_modules.await?;
+    let graph = graph.read().await?;
+
+    let module_side_effects = graph
+        .enumerate_nodes()
+        .map(async |(_, node)| {
+            Ok(match node {
+                super::SingleModuleGraphNode::Module(module) => {
+                    // This turbo task always has a cache hit since it is called when building the
+                    // module graph. we could consider moving this information
+                    // into to the module graph, but then changes would invalidate the whole graph.
+                    (*module, *module.side_effects().await?)
+                }
+                super::SingleModuleGraphNode::VisitedModule { idx: _, module } => (
+                    *module,
+                    if parent_side_effect_free_modules.contains(module) {
+                        ModuleSideEffects::SideEffectFree
+                    } else {
+                        ModuleSideEffects::SideEffectful
+                    },
+                ),
+            })
+        })
+        .try_join()
+        .await?
+        .into_iter()
+        .collect::<FxHashMap<_, _>>();
+
+    // Modules are categorized as side-effectful, locally side effect free and side effect free.
+    // So we are really just interested in determining what modules that are locally side effect
+    // free. logically we want to start at all such modules are determine if their transitive
+    // dependencies are side effect free.
+
+    let mut locally_side_effect_free_modules_that_have_side_effects = FxHashSet::default();
+    graph.traverse_edges_from_entries_dfs_reversed(
+        // Start from all the side effectful nodes
+        module_side_effects.iter().filter_map(|(m, e)| {
+            if *e == ModuleSideEffects::SideEffectful {
+                Some(*m)
+            } else {
+                None
+            }
+        }),
+        &mut (),
+        // child is a previously visited module that we know is side effectful
+        |child, _parent, _s| {
+            Ok(if let Some((child_module, _edge)) = child {
+                match module_side_effects.get(&child_module).unwrap() {
+                    ModuleSideEffects::SideEffectful | ModuleSideEffects::SideEffectFree => {
+                        // We have either already seen this or don't want to follow it
+                        GraphTraversalAction::Exclude
+                    }
+                    ModuleSideEffects::ModuleEvaluationIsSideEffectFree => {
+                        // this module is side effect free locally but must depend on something
+                        // effectful so it to is effectful
+                        locally_side_effect_free_modules_that_have_side_effects
+                            .insert(child_module);
+                        GraphTraversalAction::Continue
+                    }
+                }
```

Comment:
```suggestion
                match module_side_effects.get(&_parent).unwrap() {
                    ModuleSideEffects::SideEffectful | ModuleSideEffects::SideEffectFree => {
                        // We have either already seen this or don't want to follow it
                        GraphTraversalAction::Exclude
                    }
                    ModuleSideEffects::ModuleEvaluationIsSideEffectFree => {
                        // this module is side effect free locally but must depend on something
                        // effectful so it to is effectful
                        locally_side_effect_free_modules_that_have_side_effects
                            .insert(_parent);
                        GraphTraversalAction::Continue
                    }
                }
```

Not super sure, but I feel like this is incorrect here.

* We are coming from `child` (which was already visited and has side effects).
* And visiting `parent`.
* If `parent` is `ModuleEvaluationIsSideEffectFree` it has a child with side effects and is therefore also side effect full.

Reply:
correct! good catch